### PR TITLE
Add filtering to Account Objects (RIPD-868)

### PR DESCRIPTION
**Pull request by miguelportilla**
The 'type' field can be optionally specified to retrieve account objects of that type.

Reviewers: @nbougalis @rec

### CODE REVIEW

**Comment by rec** [comment]
Build failures are real!  https://travis-ci.org/ripple/rippled/jobs/60573888

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 1]
```diff
@@ -1132,61 +1132,85 @@ int NetworkOPsImp::findTransactionsByDestination (
...
+    }
+
+    auto dir = ledger->getDirNode (dirIndex);
+    if (dir == nullptr)
+        return false;
 
-        auto const hintIndex = getDirNodeIndex (rootIndex, hint);
-        SLE::pointer hintDir = lpLedger->getSLEi (hintIndex);
-        if (hintDir != nullptr)
+    std::uint32_t i {0};
```
Comment (rec) [inline src/ripple/app/misc/NetworkOPs.cpp]: "It's a trap!"  `i` is almost always a for loop variable.  Why not call it `objectCount`? 

For that matter, why not rename `limit` to `maxObjectCount`?

Comment (rec) [inline src/ripple/app/misc/NetworkOPs.cpp]: Why not just say `std::uint32_t i = 0;` - less to think about for a C++ programmer?

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 2]
```diff
@@ -1132,61 +1132,85 @@ int NetworkOPsImp::findTransactionsByDestination (
...
+    auto& jvObjects =
```
Comment (rec) [inline src/ripple/app/misc/NetworkOPs.cpp]: Assigning and creating a new variable at the same time is a little unexpected.

Why not

    auto& jvObjects = jvResult[jss::account_objects];
    jvObjects.resize (0):

File: src/ripple/app/misc/NetworkOPs.cpp (hunk 1 above)
Comment (vinniefalco) [inline src/ripple/app/misc/NetworkOPs.cpp]: Prefer `limit to `maxObjectCount` or its equivalent `maxNumberOfObjectsToReturnFromRPC`

**Comment by rec** [comment]
Can you do something to fix the travis pls?

**Comment by miguelportilla** [comment]
@rec Working on it. Currently only have mc and clang, working on getting my gcc on. Please stand by.

**Comment by miguelportilla** [comment]
@rec Fixed.

**Comment by rec** [comment]
> Replacing bool with auto is going slightly overboard, imo. I wouldn't hold up the PR over this, just a comment as I'm scrolling through.

I adopted C++11 in my personal codebase over two years ago.  You could see that I didn't use `auto` that much as I was suspicious of it.

Fast forward to today.  Now I think you should use `auto` every time _unless_ there's a reason not to (maybe not a huge reason, but a reason). A non-use of auto should mean, "The explicit type is important here."

Some writers think you should use `auto` for _every_ local variable - i.e. `auto foo = std::uint32_t (0);`  When I first read it I laughed but I honestly now am not sure I have a good reason against that.

**Comment by rec** [comment]
We only "needlessly allocate variables" in the rare error case, which should never happen in normal operation.

I don't agree that the longer code is easier to read.  For example, it really isn't clear that you're making the same call to `context.netOps.getAccountObjects()` twice in the original code.  And all the duplicated boilerplate hides the actual operation of the code.  

"Don't Repeat Yourself."

**Comment by miguelportilla** [comment]
@rec We would needlessly allocate whenever a marker would not be specified, which is most of the time. The 'easier' to read bit was referring to the the conditions separately listed, which is without doubt easier to read and debug. Not to worry though I've managed to confirm the second branch can never fail (as I had suspected) and simplified the code. One stone killed several birds.

**Comment by miguelportilla** [comment]
I moved the getAccountsObjects method from NetworkOps to Ledger, a more appropriate home.

**Comment by vinniefalco** [comment]
@miguelportilla In that case can't `getAccountObjects` be a free function?

**Comment by rec** [comment]
>  One stone killed several birds.

Love it!

**Comment by miguelportilla** [comment]
@vinniefalco It can be.

**Comment by vinniefalco** [comment]
@miguelportilla In that case we definitely should simply make it a free function. Perhaps part of the RPC module in its own header, since that's the only code that needs it.

File: src/ripple/app/ledger/Ledger.cpp [hunk 3]
```diff
@@ -1253,6 +1253,85 @@ bool Ledger::visitAccountItems (
...
 }
 
+bool Ledger::getAccountObjects (Account const& account,
+    LedgerEntryType const type, uint256 dirIndex, uint256 const& entryIndex,
+    std::uint32_t const limit, Json::Value& jvResult) const
+{
+    auto const rootDirIndex = getOwnerDirIndex (account);
+    auto found = false;
+
+    if (dirIndex.isZero ())
```
Comment (nbougalis) [inline src/ripple/app/ledger/Ledger.cpp]: `base_uint::isZero` and `base_uint::isNonZero` are deprecated.

Recommend:
    if (dirIndex == zero)

and
    if (dirIndex != zero)

File: src/ripple/rpc/handlers/AccountObjects.cpp [hunk 4]
```diff
@@ -61,10 +67,39 @@ Json::Value doAccountObjects (RPC::Context& context)
...
+            { "ticket", ltTICKET }
+        };
+
+        auto const& p = params[jss::type];
+        if (! p.isString ())
+            return RPC::expected_field_error (jss::type, "string");
+
+        auto const filter = p.asString ();
+        auto iter = std::find_if (types.begin (), types.end (),
+            [&](decltype (types.front ())& t) { return t.first == filter; });
```
Comment (nbougalis) [inline src/ripple/rpc/handlers/AccountObjects.cpp]: You could just capture `filter` here, since you use nothing else: `[&filter]`

**Comment by nbougalis** [comment]
LGTM :+1:

**Comment by vinniefalco** [comment]
Squash please

**Pull request closed without merging**